substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/parallel/ParallelGCImpl.java

```diff
@@ -0,0 +1,258 @@
+            FastThreadLocalFactory.createWord("ParallelGCImpl.scannedChunkTL");
+
+    public static final VMMutex mutex = new VMMutex("ParallelGCImpl");
+    private final VMCondition seqPhase = new VMCondition(mutex);
+    private final VMCondition parPhase = new VMCondition(mutex);
+
+    private ChunkBuffer buffer;
+    private volatile boolean inParallelPhase;
+
+    @Fold
+    public static ParallelGCImpl singleton() {
+        return (ParallelGCImpl) ImageSingletons.lookup(ParallelGC.class);
+    }
+
+    public static boolean isInParallelPhase() {
+        return singleton().inParallelPhase;
+    }
+
+    public static void waitForIdle() {
+        singleton().waitForIdleImpl();
+    }
+
+    public static AlignedHeapChunk.AlignedHeader getThreadLocalScannedChunk() {
+        return scannedChunkTL.get();
+    }
+
+    public static AlignedHeapChunk.AlignedHeader getThreadLocalChunk() {
+        return allocChunkTL.get();
+    }
+
+    public static void setThreadLocalChunk(AlignedHeapChunk.AlignedHeader chunk) {
+        allocChunkTL.set(chunk);
+    }
+
+    public void push(Pointer ptr) {
+        buffer.push(ptr);
+    }
+
+    @Override
+    public void startWorkerThreadsImpl() {
+        int hubOffset = ConfigurationValues.getObjectLayout().getHubOffset();
+        VMError.guarantee(hubOffset == 0, "hub offset must be 0");
+
+        // Allocate buffer large enough to store maximum possible number of heap chunks
+        long maxHeapSize = GCImpl.getPolicy().getMaximumHeapSize().rawValue();
+        long alignedChunkSize = HeapParameters.getAlignedHeapChunkSize().rawValue();
+        long unalignedChunkSize = HeapParameters.getLargeArrayThreshold().rawValue();
+        long maxChunks = maxHeapSize / Math.min(alignedChunkSize, unalignedChunkSize) + 1;
+        buffer = new ChunkBuffer((int) maxChunks, ConfigurationValues.getObjectLayout().getReferenceSize());
+
+        workerCount = getWorkerCount();
+        allWorkersBusy = ~(-1 << workerCount) & (-1 << 1);
+        // We reuse the gc thread for worker #0
+        for (int i = 1; i < workerCount; i++) {
+            startWorkerThread(i);
+        }
+    }
+
+    private int getWorkerCount() {
+        int setting = ParallelGCOptions.ParallelGCWorkers.getValue();
+        int workers = setting > 0 ? setting : getDefaultWorkerCount();
+        workers = Math.min(workers, MAX_WORKERS_COUNT);
+        verboseGCLog().string("[Number of ParallelGC workers: ").unsigned(workers).string("]").newline();
+        return workers;
+    }
+
+    private int getDefaultWorkerCount() {
+        // Adapted from Hotspot, see WorkerPolicy::nof_parallel_worker_threads()
+        int cpus = Jvm.JVM_ActiveProcessorCount();
+        return cpus <= 8 ? cpus : 8 + (cpus - 8) * 5 / 8;
+    }
+
+    private void startWorkerThread(int n) {
+        Thread t = new Thread(() -> {
+                VMThreads.SafepointBehavior.markThreadAsCrashed();
```

Comment:
Are there any examples of unattached threads usage in the CE code base? I guess I need something to learn from

Reply:
I spent some more time on that. Unfortunately, we can't use unattached threads as they would cause too many issues for the Native Image GC code. So, it will be necessary to use normal Java threads, similar to what you are already doing at the moment.

However, using Java threads also has certain issues. I will try to come up with a design for a proper solution that really works for all relevant use-cases. Until then, I would suggest that you keep your code as it is.